src/Features/Core/Portable/Snippets/RoslynLSPSnippetConverter.cs

```diff
@@ -0,0 +1,174 @@
+        {
+            var textChangeStart = textChange.Span.Start;
+            var textChangeText = textChange.NewText!;
+            using var _ = PooledStringBuilder.GetInstance(out var lspSnippetString);
+            var map = GetMapOfSpanStartsToLSPStringItem(placeholders, textChangeStart);
+
+            for (var i = 0; i < textChangeText.Length + 1;)
+            {
+                if (i == caretPosition - textChangeStart)
+                {
+                    lspSnippetString.Append("$0");
+
+                    // Special case for cursor position since they will occur between positions
+                    // so we still wants to insert the character following the cursor position.
+                    if (i < textChangeText.Length)
+                    {
+                        lspSnippetString.Append(textChangeText[i]);
+                    }
+
+                    i++;
+                }
+
+                var (str, strLength) = GetStringInPosition(map, position: i);
+                if (str.IsEmpty())
+                {
+                    if (i < textChangeText.Length)
+                    {
+                        lspSnippetString.Append(textChangeText[i]);
+                        i++;
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    lspSnippetString.Append(str);
+                    i += strLength;
+                }
+            }
+
+            return lspSnippetString.ToString();
+        }
+
+        /// <summary>
+        /// Preprocesses the list of placeholders into a dictionary that maps the insertion position
+        /// in the string to the placeholder's identifier and the number associated with it.
+        /// </summary>
+        private static Dictionary<int, (string identifier, int priority)> GetMapOfSpanStartsToLSPStringItem(ImmutableArray<SnippetPlaceholder> placeholders, int textChangeStart)
+        {
+            var map = new Dictionary<int, (string, int)>();
+
+            for (var i = 0; i < placeholders.Length; i++)
+            {
+                var placeholder = placeholders[i];
+                foreach (var position in placeholder.PlaceHolderPositions)
+                {
+                    map.Add(position - textChangeStart, (placeholder.Identifier, i + 1));
+                }
+            }
+
+            return map;
+        }
+
+        /// <summary>
+        /// Tries to see if a value exists at that position in the map, and if so it
+        /// generates a string that is LSP formatted as well as passes back the length
+        /// of the identifier so that it can skip forward in the string.
+        /// </summary>
+        private static (string str, int strLength) GetStringInPosition(Dictionary<int, (string identifier, int priority)> map, int position)
```

Comment:
only called in one place, and that claler basically has to redo the if-check.  consider just inlining this.